File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    const graph = Array(n+1).fill().map(()=>[])
    
    for (const [v1,v2] of wires){
        graph[v1].push(v2)
        graph[v2].push(v1)
    }

    let answer = Infinity
    for (const [v1,v2] of wires){
        const g = [...graph]
        g[v1] = g[v1].filter(v => v !== v2)
        g[v2] = g[v2].filter(v => v !== v1)
        let visited = Array(n+1).fill(false)
        let count = 0
        dfs(1, g, visited,0)
        count = visited.filter(v=>v).length
        answer = Math.min(answer, Math.abs(n - count - count))
    }
    return answer;
}

function dfs(start, graph, visited){
    visited[start] = true
    for (let next of graph[start]){
        if (!visited[next]) {
            visited[next] = true
            dfs(next, graph, visited)
        
        }
    }
    
}

File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    graph = {}

    for u,v in wires:
        if u in graph:
            graph[u].append(v)
        else:
            graph[u] = [v]
        if v in graph:
            graph[v].append(u)
        else:
            graph[v] = [u]

    def dfs(i, graph, visited):
        visited.add(i)
        count = 1
        for v in graph[i]:
            if not v in visited:
                count += dfs(v, graph, visited)
        return count
    
    answer = n
    
    for u, v in wires:
        visited = set()
        graph[u].remove(v) 
        graph[v].remove(u) 
        count = dfs(1,graph,visited)
        answer = min(answer, abs(n - count - count))
        graph[u].append(v) 
        graph[v].append(u) 

    return answer